ISPN-14872 Include RESP in AbstractAuthorization tests

jabolina opened the pull request:
https://issues.redhat.com/browse/ISPN-14872
https://issues.redhat.com/browse/ISPN-14895

Besides adding the tests with the server authorization, this also includes a way to authenticate using a certificate. The only thing not covered is Kerberos, which is skipped.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:
```diff
@@ -38,12 +39,29 @@ public CompletionStage<Boolean> performAuth(ChannelHandlerContext ctx, byte[] us
       if (authenticator == null) {
-         return CompletableFutures.booleanStage(handleAuthResponse(ctx, null));
+         boolean success = handleAuthResponse(ctx, null);
+         return success ? CompletableFutures.completedTrue() : CompletableFutures.completedFalse();
```
The original code was better :)

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:
```diff
@@ -38,12 +39,29 @@ public CompletionStage<Boolean> performAuth(ChannelHandlerContext ctx, byte[] us
+      }
+
+      CompletionStage<Subject> authentication;
+      if (username == null && password == null) {
```
Probably you should also check `authenticator.isClientCertAuthEnabled()` here

tristantarrant commented on server/runtime/src/main/java/org/infinispan/server/configuration/endpoint/EndpointConfigurationBuilder.java:
```diff
@@ -272,17 +274,31 @@ public static void enableImplicitAuthentication(SecurityConfiguration security,
             // The connector was added implicitly, but the security realm cannot support it. Remove it.
             return null;
          } else {
             throw Server.log.respEndpointRequiresRealmWithPassword();
```
Change the log to also say "or a truststore realm"

tristantarrant commented on server/runtime/src/main/java/org/infinispan/server/security/RespClientCertAuthenticator.java:
```diff
@@ -0,0 +1,53 @@
+
+
+   public CompletionStage<Subject> clientCertAuth(Channel channel) throws SaslException {
+      SaslAuthenticator sap = new SimpleSaslAuthenticator();
```
You should not be using the `SimpleSaslAuthenticator` but rather rely on the Elytron SASL authenticator which has been configured with a `TrustStoreRealm` and therefore can validate the client certificates against the trust store.

tristantarrant reviewed (changes requested):
I would really like to split `AbstractAuthorization` into per-protocol classes and turn the various `*AuthorizationIT` classes into `Suite`s.

jabolina commented:
Applied all the suggestions. Since we have the command `INFO`, I added a test for it, meaning slight changes, storing the subject with an attribute key and the permission check.

Now, I'll focus on trying to split the tests per protocol. But the other parts are ready for review again.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java:
```diff
@@ -76,10 +81,7 @@ public CompletionStage<RespRequestHandler> delegate(ChannelHandlerContext ctx,
-         authorizationManager.checkPermission(authorizationPermission);
-      }
+   public void checkPermission(Subject subject, AuthorizationPermission authorizationPermission) {
+      SecurityActions.checkPermission(respServer.getCacheManager(), subject, authorizationPermission);
```
This code is wrong. 
Some authz checks need to be done against the cache permissions, while others may need to be done using the global permissions.

jabolina commented:
Fixed the previous issues. Some resources were leaking after the test was finished. But now Lettuce threads are leaking.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:
```diff
@@ -5,14 +5,17 @@
+import io.netty.util.AttributeKey;
 
 public class Resp3AuthHandler extends CacheRespRequestHandler {
+   static final AttributeKey<Subject> SUBJECT_ATTRIBUTE_KEY = AttributeKey.newInstance("subject");
```
@jabolina I'm adding the Subject to the `ConnectionMetadata` in https://github.com/infinispan/infinispan/pull/10964

jabolina commented:
Updated removing the Subject attr key. The last CI run had a [failure in the REST backup test](https://ci.infinispan.org/job/Infinispan/job/PR-10961/8/testReport/org.infinispan.server.security.authorization/RESTAuthorizationTest/testRestAdminsMustAccessBackupsAndRestores/).

This was triggered by running the RESP tests first, and since the `respCache` would have heterogeneous entries, plain `byte[]` and multimap objects, the test would fail. I added a little workaround checking the metadata, but that's until we establish a way to verify the entries properly for data structures.

jabolina commented:
Looks like I broke everything again.

jabolina commented:
Well, the thread dump had almost 300 Ok Http threads, and the CI logs have some hints about that, too. I took some time removing all usages of `sync` in the REST tests, which could leak some resources. Let's see how it goes.

jabolina commented:
Looks like the failures are unrelated.

tristantarrant commented:
Merged, thanks
